packages/test/test-end-to-end-tests/src/test/attachLifecycle.spec.ts

```diff
@@ -0,0 +1,157 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { strict as assert } from "assert";
+import { requestFluidObject } from "@fluidframework/runtime-utils";
+import {
+    createLoader,
+    ITestFluidObject,
+    timeoutPromise,
+} from "@fluidframework/test-utils";
+
+import {generatePairwiseOptions} from "@fluidframework/test-pairwise-generator";
+import { describeFullCompat } from "@fluidframework/test-version-utils";
+import { IResolvedUrl } from "@fluidframework/driver-definitions";
+import { ISharedMap, IValueChanged, SharedMap } from "@fluidframework/map";
+import { IFluidHandle } from "@fluidframework/core-interfaces";
+import { AttachState } from "@fluidframework/container-definitions";
+
+//these points are all after creation of all object at which any object can be attached
+const sharedPoints = [3,4,5];
+
+const testConfigs =
+    generatePairwiseOptions({
+        containerAttachPoint:[0, ... sharedPoints],
+        datastoreAttachPoint: [1, ... sharedPoints],
+        datastoreSaveAfterAttach: [true, false],
+        ddsAttachPoint: [2, ... sharedPoints],
```

Comment:
The non-sharedPoints values for each of the *AttachPoints was a bit confusing to me without reading the code closely. Is it true that there is nothing interesting to test by including the other potential values after the objects are created? e.g. "containerAttachPoint:[0,1,2,3,4,5]"

Similarly, why do we not want to test duplicate values of attach points?

Reply:
Oh, I see dupes are allowed now.

Reply:
so 0,1, and 2 are special, because at those points the preceding things haven't even been create, so for instance at point 0 if we attach, the datastore and the map are both create on attach containers, which has implications for how we manage our internal, or at least it could. by 3,4, and 5 everything has been created, at least locally, so any attach is in the presence of all objects. you also cannot create a dds without a datastore, or a datastore without a container, that's what those points are coupled to their specific objects. this gives coverage for existence and attach state.

writing this has actually made me think adding 1 to the containerAttachPoints is interesting, attach container after datastore creation, but before dds creation. this case wouldn't have been covered another way. i don't think 2 is interesting for either datastore or container, as its the same as the shared points, as by necessity everything must have been created, at least locally by the time 2 is over.

for duplicates, i've removed that, see my reply to marks comment above